ledger-core/network/nds/msgdelivery/controller_parallelism_test.go

```diff
@@ -0,0 +1,273 @@
+		require.NoError(t, err)
+	}
+
+	<-await
+
+	for i := 0; i < numberMessage; i++ {
+		err := srv3.service.ShipTo(srv1.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("%d", i)}})
+		require.NoError(t, err)
+	}
+
+	select {
+	case res, ok := <-resultsOnServ2:
+		require.True(t, ok)
+		t.Fatalf("unexpected string received:%s", res)
+	default:
+		// channel is empty, this okay
+	}
+
+	checker := func(ch chan string) {
+		res := make(map[string]string)
+
+		for i := 0; i < numberMessage; i++ {
+			val := <-ch
+
+			res[val] = val
+		}
+
+		for i := 0; i < numberMessage; i++ {
+			expVal := fmt.Sprintf("%d", i)
+			_, ok := res[expVal]
+			require.Truef(t, ok, "Not received:%s\n", expVal)
+		}
+	}
+
+	checker(resultsOnServ1)
+
+	out.stop.Store(false)
+	resume <- struct{}{}
+
+	checker(resultsOnServ2)
+}
+
+func TestParallelismReceive(t *testing.T) {
+	h := NewUnitProtoServersHolder(TestLogAdapter{t: t})
+	defer h.stop()
+
+	prf := UnitProtoServerProfile{
+		desFactory: &TestDeserializationStringFactory{},
+	}
+
+	prf1 := prf
+	prf2 := prf
+	prf3 := prf
+
+	pause := make(chan struct{})
+
+	numberMessage := 1000
+
+	receivedFrom1 := make(chan string, numberMessage)
+	receivedFrom3 := make(chan string, numberMessage)
+
+	first := atomickit.NewBool(true)
+
+	receiver2 := func(a ReturnAddress, done nwapi.PayloadCompleteness, val interface{}) error {
+		require.True(t, bool(done))
+		s := val.(fmt.Stringer).String()
+
+		switch {
+		case strings.HasPrefix(s, "1"):
+			if first.Load() {
+				first.Flip()
+				pause <- struct{}{}
+				<-pause
+			}
+
+			receivedFrom1 <- s
+		case strings.HasPrefix(s, "3"):
+			receivedFrom3 <- s
+		}
+
+		return nil
+	}
+
+	srv1, err := h.createServiceWithProfile(&prf1, noopReceiver)
+	require.NoError(t, err)
+	srv2, err := h.createServiceWithProfile(&prf2, receiver2)
+	require.NoError(t, err)
+	srv3, err := h.createServiceWithProfile(&prf3, noopReceiver)
+	require.NoError(t, err)
+
+	err = srv1.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("1%d", 0)}})
+	require.NoError(t, err)
+
+	<-pause
+
+	for i := 1; i < numberMessage; i++ {
+		err := srv1.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("1%d", i)}})
+		require.NoError(t, err)
+	}
+
+	for i := 0; i < numberMessage; i++ {
+		err := srv3.service.ShipTo(srv2.directAddress(), Shipment{Head: &TestString{fmt.Sprintf("3%d", i)}})
+		require.NoError(t, err)
+	}
+
+	checker := func(from string, startFrom int, ch chan string) {
+		res := make(map[int]string)
+
+		for i := startFrom; i < numberMessage; i++ {
+			val := <-ch
+
+			val0 := strings.TrimPrefix(val, from)
+
+			parseInt, e := strconv.ParseInt(val0, 10, 32)
+			require.NoError(t, e)
+			res[int(parseInt)] = val
+		}
+
+		for i := startFrom; i < numberMessage; i++ {
+			val := res[i]
+			require.Equal(t, fmt.Sprintf("%s%d", from, i), val)
+		}
+	}
+
+	checker("1", 1, receivedFrom1)
+	checker("3", 0, receivedFrom3)
+
+	pause <- struct{}{}
```

Comment:
better to use close()

Reply:
close for what? We should not stop transport here.

Reply:
it was about (pause)

Reply:
Fixed